### ISSUE TITLE
Why not Masked Prompt Tokens in SFT Loss Calculation for LLMs?

### ISSUE THREAD

**Issue by ParadoxZW** [huggingface/trl#3279]
I noticed that in [sft.py](https://github.com/huggingface/trl/blob/main/trl/scripts/sft.py), when performing supervised fine-tuning (SFT) of an LLM on an instruction-following dataset, the next-token prediction loss for the prompt tokens is not masked out (e.g., by setting the label to -200, which would allow the loss function to automatically ignore those tokens). I would like to ask: Is this a common practice in SFT? Does computing the autoregressive loss on the prompt tokens along with the response tokens have (almost) no impact on model performance (so making implementation simpler), or does it provide additional performance benefits?

P.S. I looked into this issue and found that the `tokenizer.apply_chat_template` interface has a parameter `return_assistant_tokens_mask`, which, when used with a specific Jinja `chat_template`, can conveniently output a labels tensor with -200 for masked tokens. Something like follows,
```
chat_template = """\
{%- if messages[0]['role'] == 'system' %}
    {{- '<|im_start|>system\\n' + messages[0]['content'] + '<|im_end|>\\n' }}
{%- else %}
    {{- '<|im_start|>system\\nYou are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>\\n' }}
{%- endif %}
{%- for message in messages %}
    {%- if (message.role == "user") or (message.role == "system" and not loop.first)%}
        {{- '<|im_start|>' + message.role + '\\n' + message.content + '<|im_end|>' + '\\n' }}
    {%- elif message.role == "assistant" %}
        {{- '<|im_start|>' + message.role + '\\n'}}
        {%- generation %}
        {{- message.content + '<|im_end|>'}}
        {%- endgeneration %}
        {{- '\\n' }}
    {%- endif %}
{%- endfor %}
{%- if add_generation_prompt %}
    {{- '<|im_start|>assistant\\n' }}
{%- endif %}
"""

def formatting_func(
    example,
    tokenizer,
):
    output = tokenizer.apply_chat_template(
        example["messages"], tokenize=True, return_dict=True, return_tensors="pt",
        chat_template=chat_template, add_generation_prompt=False, truncation=True,
        return_assistant_tokens_mask=True,
    )
    labels = output.input_ids.clone()
    labels[output.input_ids==tokenizer.pad_token_id] = -100
    labels[output.assistant_masks==0] = -100  # 非 assistant 部分设为 -100
    data = {
        "input_ids": output.input_ids[0].tolist(),
        "labels": labels[0].tolist()
    }
    return data

raw_dataset = load_dataset("./ultrachat_200k")
train_dataset = raw_dataset["train_sft"]
test_dataset = raw_dataset["test_sft"]
column_names = list(train_dataset.features)

processed_train_dataset = train_dataset.map(
    formatting_func,
    fn_kwargs={"tokenizer": tokenizer},
    num_proc=10,
    remove_columns=column_names,
    desc="Applying chat template to train_sft",
)
```

**Comment by benjamin-marie** [comment]
Do you mean training on completions only?
There is this:
https://huggingface.co/docs/trl/en/sft_trainer#train-on-completions-only

**Comment by ParadoxZW** [comment]
> Do you mean training on completions only? There is this: https://huggingface.co/docs/trl/en/sft_trainer#train-on-completions-only

Thanks! It seems `DataCollatorForCompletionOnlyLM` is a convenient interface to implement what I want. I haven't know this interface before. But do you think my implementation can also work properly?

I also notice

> Note that this works only in the case when packing=False.

while my implementation way can set `packing=True`.

Do you think I can PR my implementation to the TRL repo?

**Comment by qgallouedec** [comment]
@ParadoxZW I've been wanting to make the completions-only training simpler for a while, but haven't had the time to get around to it yet. I like your approach and the use of `return_assistant_tokens_mask`. Ideally, we'd use it in the `trl/data_utils.py` functions. If you feel like opening a PR it would be very welcome!

**Comment by ParadoxZW** [comment]
@qgallouedec Sure, I'd love to do that.

I've spent days to modify my training script and to examine if the implementations works as expected. Following is my main modification:

```
# <omit>

##################
# Data Processing
##################
chat_template = """\
{%- for message in messages %}
    {%- if (message.role == "user") %}
        {{- '<|im_start|>' + message.role + '\\n' + message.content + '<|im_end|>' + '\\n' }}
    {%- elif message.role == "assistant" %}
        {{- '<|im_start|>' + message.role + '\\n'}}
        {%- generation %}
        {{- message.content + '<|im_end|>'}}
        {%- endgeneration %}
        {{- '\\n' }}
    {%- endif %}
{%- endfor %}
{%- if add_generation_prompt %}
    {{- '<|im_start|>assistant\\n' }}
{%- endif %}
"""

def formatting_func(
    example,
    tokenizer,
):
    output = tokenizer.apply_chat_template(
        example["messages"], tokenize=True, return_dict=True, return_tensors="pt",
        chat_template=chat_template, add_generation_prompt=False, truncation=True,
        return_assistant_tokens_mask=True,
    )
    labels = output.input_ids.clone()
    labels[output.input_ids==tokenizer.pad_token_id] = -100
    labels[output.assistant_masks==0] = -100  # assign -100 to non-assistant part
    data = {
        "input_ids": output.input_ids[0].tolist(),
        "attention_mask": output.attention_mask[0].tolist(),
        "labels": labels[0].tolist()
    }
    return data

raw_dataset = load_dataset("./ultrachat_200k")
train_dataset = raw_dataset["train_sft"]
column_names = list(train_dataset.features)

processed_train_dataset = train_dataset.map(
    formatting_func, fn_kwargs={"tokenizer": tokenizer},
    remove_columns=column_names, num_proc=10,
    desc="Applying chat template to train_sft",
)

# do packing by ourselves.
def pack_examples_without_fragmentation(
    examples: dict[str, list[list]], seq_length: int
) -> dict[str, list[list]]:
    """
    Pack examples into chunks of size less than `seq_length`. The difference with `pack_examples` is 
    that this function does not continue to concatenate a new sample if the packed length will be
    greater than `seq_length`, but instead starts a new pack.
    """
    new_examples = {}
    for k in examples.keys():
        first_data = examples[k][0]
        new_examples[k] = [first_data[:seq_length]]
        for i in range(len(examples[k])):
            data = examples[k][i]
            last_pack = new_examples[k][-1]
            if len(last_pack) + len(data) <= seq_length:
                new_examples[k][-1] = last_pack + data
            else:
                new_examples[k].append(data)
    return new_examples

processed_train_dataset = processed_train_dataset.map(
    pack_examples_without_fragmentation, 
    batched=True, fn_kwargs={"seq_length": max_seq_length},
    new_fingerprint="packing_2",
)



###########
# Training
###########
def length_variable_data_collator(batch):
    max_sample_length = max([len(x["input_ids"]) for x in batch])
    input_ids = []
    attention_mask = []
    labels = []
    for i in range(len(batch)):
        input_ids.append(batch[i]["input_ids"] + [tokenizer.pad_token_id] * (max_sample_length - len(batch[i]["input_ids"])))
        attention_mask.append(batch[i]["attention_mask"] + [0] * (max_sample_length - len(batch[i]["attention_mask"])))
        labels.append(batch[i]["labels"] + [-100] * (max_sample_length - len(batch[i]["labels"])))

    return {
        "input_ids": torch.tensor(input_ids),
        "attention_mask": torch.tensor(attention_mask),
        "labels": torch.tensor(labels)
    }

trainer = SFTTrainer(
    model=model,
    args=train_conf, # train_conf["packing"]=False, because We do packing by ourselves.
    data_collator=length_variable_data_collator,
    train_dataset=processed_train_dataset,
)

# <omit>
```

To conclude, I've done following things:
1. use `return_assistant_tokens_mask` in the function `formatting_func` to realize completions-only training
2. do packing by my own. `pack_examples_without_fragmentation` will concatenate samples without break them (the original pack function in `trl` will break samples)
3. since packed samples have variable length, I implement a new collector `length_variable_data_collator`

Do you think I can move above modifications into the main branch of `trl`? Or do you have any suggestions, like which part is unnecessary or need to be improved, and how to merge my code to `trl` in an elegant way?

I am looking forward to your reply and wanna commit a PR as soon as possible.

**Comment by albertvillanova** [comment]
Hey @ParadoxZW, sorry for the very slow reply — we're catching up on the backlog. 

Good news: exactly what you proposed got built. `SFTTrainer` now supports `assistant_only_loss=True` in `SFTConfig`, which uses `return_assistant_tokens_mask` + `{% generation %}` chat-template markers to mask everything but the assistant turns (https://github.com/huggingface/trl/blob/0ed4ac7fa03bf388a94566f1b1f71e71445ca6e7/trl/trainer/sft_config.py#L260-L269), and `completion_only_loss` does the same for prompt-completion datasets. Both work with packing now (the old `DataCollatorForCompletionOnlyLM` `packing=False` limitation is gone), and the non-fragmenting packing you implemented by hand is now the default `packing_strategy="bfd"`. 

To your original question: masking the prompt is the standard choice for instruction tuning, and it's the default for prompt-completion data. 

Closing as resolved since the feature landed — happy to reopen if anything's missing!
